File: apps/android/app/src/main/java/com/voiid/app/onboarding/VerifiedScreen.kt
package com.voiid.app.onboarding

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.EaseOut
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.scale
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathMeasure
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.voiid.app.ui.components.LocalVoiidHaptics
import com.voiid.app.ui.components.reduceMotionEnabled
import com.voiid.app.ui.theme.VoiidColor
import com.voiid.app.ui.theme.VoiidFont
import com.voiid.app.ui.theme.VoiidPalette
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * The success moment, after the code is accepted. Port of iOS `VerifiedScreen`.
 *
 * Onboarding is otherwise deliberately still; this screen is the exception because it happens
 * ONCE per account (the rare/first-time tier). It also has a job beyond delight: verification is
 * when login and E2E bootstrap are still settling, so the animation converts that wait into
 * confirmation.
 *
 * Choreography: ring draws → tick strokes on (overlapping the ring's tail) → both settle → copy
 * fades up. Sequenced so the eye sees the mark being MADE. Reduced motion presents the finished
 * mark instantly and keeps only the copy fade — a hard cut is its own jarring.
 */
@Composable
fun VerifiedScreen(
    onFinished: () -> Unit,
    holdAfterMs: Long = 1_100,
) {
    val haptics = LocalVoiidHaptics.current
    val reduceMotion = reduceMotionEnabled()
    // Resolve theme colours once at composition — the token getters are @Composable and cannot
    // be read inside draw lambdas.
    val groundColor = VoiidColor.background
    val primary = VoiidColor.primary
    val textPrimary = VoiidColor.textPrimary
    val textSecondary = VoiidColor.textSecondary

    // 0→1, draws the ring.
    val ring = remember { Animatable(0f) }
    // 0→1, strokes the tick.
    val tick = remember { Animatable(0f) }
    // The settle pop once the tick lands.
    val markScale = remember { Animatable(0.86f) }
    // One soft halo that expands and fades as the tick completes.
    val haloScale = remember { Animatable(0.9f) }
    val haloOpacity = remember { Animatable(0f) }
    val copyAlpha = remember { Animatable(0f) }

    LaunchedEffect(Unit) {
        if (reduceMotion) {
            // Present, but not drawn. See the header comment.
            ring.snapTo(1f)
            tick.snapTo(1f)
            markScale.snapTo(1f)
            haloOpacity.snapTo(0f)
            copyAlpha.animateTo(1f, tween(durationMillis = 300, easing = EaseOut))
        } else {
            coroutineScope {
                // 1. The ring sweeps closed. Ease-out settles at the join so the circle appears
                //    to close rather than to stop.
                launch { ring.animateTo(1f, tween(durationMillis = 450, easing = EaseOut)) }
                delay(260)

                // 2. The tick strokes on, overlapping the ring's tail — waiting for the ring to
                //    finish first would put a dead beat in the middle of the moment.
                launch { tick.animateTo(1f, tween(durationMillis = 300, easing = EaseOut)) }
                haptics.success()

                // 3. Everything settles together. Low bounce reads as confidence, not a toy.
                launch {
                    markScale.animateTo(
                        1f,
                        spring(dampingRatio = 0.72f, stiffness = Spring.StiffnessLow),
                    )
                }

                // Start visible FIRST, then animate away — assigning the final state before the
                // animation would show nothing at all.
                haloOpacity.snapTo(0.55f)
                launch { haloScale.animateTo(1.6f, tween(durationMillis = 750, easing = EaseOut)) }
                launch { haloOpacity.animateTo(0f, tween(durationMillis = 750, easing = EaseOut)) }

                delay(180)
                launch { copyAlpha.animateTo(1f, tween(durationMillis = 350, easing = EaseOut)) }
            }
        }
        delay(holdAfterMs)
        onFinished()
    }

    Box(
        Modifier
            .fillMaxSize()
            .background(groundColor),
        contentAlignment = Alignment.Center,
    ) {
        // The pool of light, matching the splash's own. Static — it is a backdrop, not the event.
        Box(
            Modifier
                .fillMaxSize()
                .drawBehind {
                    drawRect(
                        Brush.radialGradient(
                            colors = listOf(
                                primary.copy(alpha = 0.13f),
                                primary.copy(alpha = 0.03f),
                                Color.Transparent,
                            ),
                            center = Offset(size.width * 0.5f, size.height * 0.42f),
                            radius = 320.dp.toPx(),
                        )
                    )
                }
        )

        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            VerifiedMark(primary, ring, tick, markScale, haloScale, haloOpacity)

            Text(
                "Verified",
                style = VoiidFont.rounded(30, FontWeight.Bold),
                color = textPrimary,
                modifier = Modifier.padding(top = 32.dp).copyFade(copyAlpha.value, reduceMotion),
            )
            Text(
                "Your number is confirmed.",
                style = VoiidFont.rounded(15),
                color = textSecondary,
                modifier = Modifier.padding(top = 8.dp).copyFade(copyAlpha.value, reduceMotion),
            )
        }
    }
}

/** Copy fades up 8pt as it appears; reduced motion fades in place. */
private fun Modifier.copyFade(progress: Float, reduceMotion: Boolean): Modifier =
    this
        .alpha(progress)
        .let { if (reduceMotion) it else it.offset(y = ((1f - progress) * 8).dp) }

/**
 * Ring (116dp, 4dp stroke), tick (54×40, 7dp stroke), halo pool (132dp).
 * Sizes and stroke weights are the iOS values exactly.
 */
@Composable
private fun VerifiedMark(
    primary: Color,
    ring: Animatable<Float, *>,
    tick: Animatable<Float, *>,
    markScale: Animatable<Float, *>,
    haloScale: Animatable<Float, *>,
    haloOpacity: Animatable<Float, *>,
) {
    val density = LocalDensity.current
    val brightStop = VoiidPalette.PrimaryDark   // #68B8BD — the lit edge stop in both themes

    Box(Modifier.size(132.dp).scale(markScale.value), contentAlignment = Alignment.Center) {
        Canvas(
            Modifier
                .size(132.dp)
                .scale(haloScale.value)
                .alpha(haloOpacity.value)
        ) {
            drawCircle(primary.copy(alpha = 0.16f))
        }

        Canvas(Modifier.size(116.dp)) {
            if (ring.value <= 0f) return@Canvas
            // Starts at 12 o'clock so the sweep closes at the top rather than reading arbitrary.
            drawArc(
                brush = Brush.linearGradient(
                    colors = listOf(brightStop, primary),
                    start = Offset(size.width * 0.25f, 0f),
                    end = Offset(size.width * 0.75f, size.height),
                ),
                startAngle = -90f,
                sweepAngle = 360f * ring.value,
                useCenter = false,
                style = Stroke(width = 4.dp.toPx(), cap = StrokeCap.Round),
            )
        }

        Canvas(Modifier.size(54.dp, 40.dp)) {
            if (tick.value <= 0f) return@Canvas
            // A checkmark as a PATH trimmed stroke-wise: one stroke, start to end.
            val path = Path().apply {
                moveTo(0f, size.height * 0.55f)
                lineTo(size.width * 0.34f, size.height)
                lineTo(size.width, 0f)
            }
            val measure = PathMeasure()
            measure.setPath(path, false)
            val partial = Path()
            measure.getSegment(0f, measure.length * tick.value, partial, true)
            drawPath(
                partial,
                primary,
                style = Stroke(
                    width = 7.dp.toPx(),
                    cap = StrokeCap.Round,
                    join = StrokeJoin.Round,
                ),
            )
        }
    }
}

File: apps/ios/Voiid/Voiid/Onboarding/VerifiedScreen.swift
//
//  VerifiedScreen.swift
//  Voiid
//
//  The success moment, after the code is accepted.
//
//  ── WHY THIS ANIMATES WHEN ALMOST NOTHING ELSE DOES ─────────────────────────────
//  Onboarding is otherwise deliberately still. This screen is the exception, and the rule that
//  allows it is FREQUENCY: a user verifies once, ever. That is the rare/first-time tier, the one
//  place a decorative animation is defensible — the same motion on a screen seen daily would be
//  a defect.
//
//  It also has a JOB beyond delight. Verification is the moment the account becomes real, and
//  the app needs a beat to do the work that follows. A success animation converts that unavoidable
//  wait into confirmation, which is the honest version of a loading state.
//
//  ── THE CHOREOGRAPHY, AND WHY IT IS IN THIS ORDER ───────────────────────────────
//  Ring draws → tick strokes on → both settle → copy fades up. Sequenced, not simultaneous:
//  everything arriving at once reads as a single image being switched on, which is exactly what
//  a confirmation should NOT feel like. The eye needs to see the mark being MADE.
//
//  The tick is a trimmed path, so it draws the way a tick is drawn — one stroke, start to end.
//  Fading or scaling a finished checkmark is the cheap version and reads as it.
//
//  ── REDUCED MOTION ──────────────────────────────────────────────────────────────
//  Honoured. It does not mean "no animation": the tick still appears and the copy still fades,
//  because a hard cut to a finished state is its own jarring. What goes is the DRAWING — the
//  path completes instantly, the ring does not sweep, nothing scales.
//

import SwiftUI

// NOTE ON THE BRAND TOKENS: `VoiidBrand.lime*` names the SLOT (fill / lit edge / lower stop),
// not the hue — the values are Tide teal since the brand moved off the lime. See OnboardingKit.

struct VerifiedScreen: View {

    /// Called once the moment has played out. The caller decides what comes next.
    var onFinished: () -> Void = {}
    /// How long to hold the finished state before continuing.
    var holdAfter: TimeInterval = 1.1

    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    /// 0→1, draws the ring.
    @State private var ringProgress: CGFloat = 0
    /// 0→1, strokes the tick.
    @State private var tickProgress: CGFloat = 0
    /// The settle pop once the tick lands.
    @State private var markScale: CGFloat = 0.86
    @State private var copyVisible = false
    /// One soft halo that expands and fades as the tick completes.
    @State private var haloScale: CGFloat = 0.9
    @State private var haloOpacity: Double = 0

    var body: some View {
        ZStack {
            VoiidBrand.ground.ignoresSafeArea()

            // The pool of light, matching the splash's own. Static — it is a backdrop, not the
            // event.
            RadialGradient(
                colors: [VoiidBrand.lime.opacity(0.13), VoiidBrand.lime.opacity(0.03), .clear],
                center: UnitPoint(x: 0.5, y: 0.42),
                startRadius: 0,
                endRadius: 320
            )
            .ignoresSafeArea()

            VStack(spacing: 0) {
                Spacer(minLength: 0)

                mark

                Text("Verified")
                    .font(VoiidFont.rounded(30, .bold))
                    .foregroundColor(VoiidBrand.text)
                    .padding(.top, VoiidSpacing.xl)
                    .opacity(copyVisible ? 1 : 0)
                    .offset(y: copyVisible ? 0 : (reduceMotion ? 0 : 8))

                Text("Your number is confirmed.")
                    .font(VoiidFont.subhead)
                    .foregroundColor(VoiidBrand.textDim)
                    .padding(.top, VoiidSpacing.sm)
                    .opacity(copyVisible ? 1 : 0)
                    .offset(y: copyVisible ? 0 : (reduceMotion ? 0 : 8))

                Spacer(minLength: 0)
            }
        }
        .preferredColorScheme(.dark)
        .navigationBarBackButtonHidden(true)
        // The moment is over — there is nothing to go back TO, because the number is verified.
        .toolbar(.hidden, for: .navigationBar)
        .accessibilityElement(children: .combine)
        .accessibilityLabel("Verified. Your number is confirmed.")
        .task { await play() }
    }

    // MARK: The mark

    private var mark: some View {
        ZStack {
            // The halo. Expands past the ring and fades — the visual equivalent of the haptic
            // that fires with it.
            Circle()
                .fill(VoiidBrand.lime.opacity(0.16))
                .frame(width: 132, height: 132)
                .scaleEffect(haloScale)
                .opacity(haloOpacity)

            // The ring, drawn rather than faded. Starts at 12 o'clock: `trim` begins at 3, so
            // without the rotation the sweep starts from the right and reads as arbitrary.
            Circle()
                .trim(from: 0, to: ringProgress)
                .stroke(
                    LinearGradient(
                        colors: [VoiidBrand.limeBright, VoiidBrand.lime],
                        startPoint: .topLeading, endPoint: .bottomTrailing
                    ),
                    style: StrokeStyle(lineWidth: 4, lineCap: .round)
                )
                .rotationEffect(.degrees(-90))
                .frame(width: 116, height: 116)

            Tick()
                .trim(from: 0, to: tickProgress)
                .stroke(
                    VoiidBrand.lime,
                    style: StrokeStyle(lineWidth: 7, lineCap: .round, lineJoin: .round)
                )
                .frame(width: 54, height: 40)
        }
        .scaleEffect(markScale)
    }

    // MARK: The sequence

    private func play() async {
        if reduceMotion {
            // Present, but not drawn. See the header.
            ringProgress = 1
            tickProgress = 1
            markScale = 1
            withAnimation(.easeOut(duration: 0.3)) { copyVisible = true }
            try? await Task.sleep(for: .seconds(holdAfter))
            onFinished()
            return
        }

        // 1. The ring sweeps closed. ease-out: fast off the mark, settling at the join, so the
        //    circle appears to close rather than to stop.
        withAnimation(.easeOut(duration: 0.45)) { ringProgress = 1 }
        try? await Task.sleep(for: .milliseconds(260))

        // 2. The tick strokes on, overlapping the ring's tail — waiting for the ring to finish
        //    first would put a dead beat in the middle of the moment.
        withAnimation(.easeOut(duration: 0.30)) { tickProgress = 1 }
        Haptics.success()

        // 3. Everything settles together. A spring, because this is the part that should feel
        //    like an object landing; bounce stays low so it reads as confidence, not as a toy.
        withAnimation(.spring(duration: 0.5, bounce: 0.28)) { markScale = 1 }

        // The halo has to be VISIBLE before it can be animated away. Setting the start value
        // after the withAnimation block — as the first version did — simply assigned the final
        // state, so the halo appeared at full strength and stayed there, a solid disc behind the
        // tick. Start first, then animate out.
        haloOpacity = 0.55
        withAnimation(.easeOut(duration: 0.75)) {
            haloScale = 1.6
            haloOpacity = 0
        }

        try? await Task.sleep(for: .milliseconds(180))
        withAnimation(.easeOut(duration: 0.35)) { copyVisible = true }

        try? await Task.sleep(for: .seconds(holdAfter))
        onFinished()
    }
}

// MARK: - The tick

/// A checkmark as a PATH, so it can be trimmed and drawn stroke-wise.
///
/// Proportional to its frame rather than fixed points, so it scales with whatever size the call
/// site gives it. The proportions are a standard tick: the short arm about a third of the long
/// one, meeting below the midline.
private struct Tick: Shape {
    func path(in rect: CGRect) -> Path {
        var p = Path()
        p.move(to: CGPoint(x: rect.minX, y: rect.midY + rect.height * 0.05))
        p.addLine(to: CGPoint(x: rect.minX + rect.width * 0.34, y: rect.maxY))
        p.addLine(to: CGPoint(x: rect.maxX, y: rect.minY))
        return p
    }
}

// MARK: - Previews

#Preview("Verified") {
    VerifiedScreen()
}
